feat: optimize database indexes for improved query performance

flemzord opened the pull request:
## 🎯 Objective

Optimize database indexes to drastically improve query performance for Point-in-Time (PIT) queries, balance lookups, and historical metadata queries.

## 🔍 Problem Analysis

The existing indexes had a mismatch between their structure and actual query patterns:

### Issues Identified

1. **Moves table**: Indexes used `accounts_seq` but queries filtered on `account_address`
2. **Metadata tables**: Indexes used `accounts_seq`/`transactions_seq` but queries used `accounts_address`/`transactions_id`
3. **Redundant index**: `accounts_volumes_idx` duplicated the PRIMARY KEY functionality

This mismatch caused:
- Full table scans on large datasets
- Poor performance on PIT queries
- Slow balance lookups
- Inefficient metadata history queries

## ✨ Solution

### Migration 41: Optimize Query Indexes

This migration introduces a **zero-downtime** index optimization strategy:

1. ✅ **CREATE** new optimized indexes FIRST (with `CONCURRENTLY`)
2. ❌ **DROP** old suboptimal indexes AFTER new ones are ready

This ensures continuous query performance during migration execution.

## 📊 Changes

### **Moves Table** (3 new indexes, 2 dropped)

#### Dropped
- ❌ `moves_post_commit_volumes` - used `accounts_seq`
- ❌ `moves_effective_post_commit_volumes` - used `accounts_seq`

#### Added
- ✅ `idx_moves_pit_insertion` on `(account_address, asset, insertion_date DESC, seq DESC)`
  - For PIT queries with insertion_date
- ✅ `idx_moves_pit_effective` on `(account_address, asset, effective_date DESC, seq DESC)`
  - For PIT queries with effective_date
- ✅ `idx_moves_account_balance` on `(account_address, effective_date DESC, seq DESC)` 
  - INCLUDE `(asset, post_commit_effective_volumes)`
  - For balance filtering queries

### **Accounts Metadata** (1 new, 1 dropped)

#### Dropped
- ❌ `accounts_metadata_revisions` - used `accounts_seq`

#### Added
- ✅ `idx_accounts_metadata_pit` on `(accounts_address, revision DESC)` 
  - INCLUDE `(metadata, date)`
  - For historical metadata lookups

### **Transactions Metadata** (1 new, 1 dropped)

#### Dropped
- ❌ `transactions_metadata_revisions` - used `transactions_seq`

#### Added
- ✅ `idx_transactions_metadata_pit` on `(transactions_id, revision DESC)` 
  - INCLUDE `(metadata, date)`
  - For historical transaction metadata lookups

### **Accounts Volumes** (cleanup)

#### Dropped
- ❌ `accounts_volumes_idx` - redundant with PRIMARY KEY `(ledger, accounts_address, asset)`

## 🎯 Affected Queries

- `resource_aggregated_balances.go`: PIT queries with `DISTINCT ON (accounts_address, asset)`
- `resource_accounts.go`: Balance filtering with `effective_date`
- `resource_accounts.go`: Account expansion volumes queries
- `resource_accounts.go`: Historical metadata lookups
- `resource_transactions.go`: Historical transaction metadata lookups

## 🚀 Performance Impact

### Expected Improvements

- 🔥 **MAJOR** improvement for Point-in-Time (PIT) queries
- 🔥 **MAJOR** improvement for balance lookups by account
- 🔥 **MAJOR** improvement for historical metadata queries
- 💾 Minor reduction in storage (redundant index removed)

### Before vs After

| Query Type | Before | After |
|------------|--------|-------|
| PIT balance lookup | Full table scan | Index-only scan |
| Historical metadata | Seq scan + filter | Index scan |
| Balance by account | Table scan | Covered index scan |

## 🔒 Migration Safety

✅ **Zero downtime deployment**
- Uses `CONCURRENTLY` flag to avoid table locks
- Creates new indexes before dropping old ones
- No data modification - only index structure changes

✅ **Rollback safe**
- Old indexes remain until new ones are fully built
- Can be safely interrupted and restarted

## 📁 Files Changed

- `internal/storage/bucket/migrations/41-optimize-query-indexes/up.sql` (66 lines)
- `internal/storage/bucket/migrations/41-optimize-query-indexes/notes.yaml` (58 lines)

## ✅ Testing Recommendations

1. Run migration on staging environment first
2. Monitor index creation progress with:
   ```sql
   SELECT * FROM pg_stat_progress_create_index;
   ```
3. Verify query plans after migration:
   ```sql
   EXPLAIN ANALYZE SELECT ...
   ```
4. Check index sizes:
   ```sql
   SELECT schemaname, tablename, indexname, pg_size_pretty(pg_relation_size(indexrelid))
   FROM pg_indexes i
   JOIN pg_class c ON c.relname = i.indexname
   WHERE schemaname = 'your_schema';
   ```

## 📝 Notes

- Migration includes comprehensive documentation in `notes.yaml`
- Index naming follows convention: `idx_<table>_<purpose>`
- All affected query patterns have been analyzed and documented

flemzord committed:
feat: optimize database indexes for query performance

Add migration 41 to optimize database indexes for improved performance:

## Changes

### Moves Table
- **DROP**: moves_post_commit_volumes, moves_effective_post_commit_volumes
  (used accounts_seq instead of account_address)
- **ADD**: idx_moves_pit_insertion for PIT queries with insertion_date
- **ADD**: idx_moves_pit_effective for PIT queries with effective_date
- **ADD**: idx_moves_account_balance for balance lookups

### Accounts Metadata
- **DROP**: accounts_metadata_revisions (used accounts_seq)
- **ADD**: idx_accounts_metadata_pit for historical metadata queries

### Transactions Metadata
- **DROP**: transactions_metadata_revisions (used transactions_seq)
- **ADD**: idx_transactions_metadata_pit for historical metadata queries

### Accounts Volumes
- **DROP**: accounts_volumes_idx (redundant with PRIMARY KEY)

## Performance Impact
- Major improvement for Point-in-Time (PIT) queries
- Major improvement for balance lookups by account
- Major improvement for historical metadata queries
- Reduced storage by removing redundant index

## Affected Query Files
- resource_aggregated_balances.go
- resource_accounts.go
- resource_transactions.go

flemzord committed:
fix: create new indexes before dropping old ones to avoid performance degradation

Reorder migration operations to:
1. Create all new optimized indexes first (with CONCURRENTLY)
2. Drop old suboptimal indexes only after new ones are ready

This ensures continuous query performance during migration execution.

flemzord committed:
refactor: remove ledger and date columns from indexes

Remove ledger and date columns from indexes as requested:
- Moves indexes: removed 'ledger' column prefix
- Accounts metadata index: removed 'ledger' and 'date' columns
- Transactions metadata index: already optimized (ledger/date not included)

This reduces index size while maintaining query performance.

gfyrag reviewed (commented):
Concurrently created indexes must be isolated in a single file.
When processing multi statement queries, PG always create a transaction.
And concurrent indexes are not compatible with transactions.

flemzord committed:
refactor: split index migrations into isolated concurrent operations

Split the single migration 41 into 5 separate migrations:
- Migration 41: Create idx_moves_pit_insertion index
- Migration 42: Create idx_moves_pit_effective index
- Migration 43: Create idx_accounts_metadata_pit index
- Migration 44: Create idx_transactions_metadata_pit index
- Migration 45: Drop all old indexes

This ensures each concurrent index creation is isolated in its own
migration, and all index drops are consolidated into a final migration.
Also removed the optional idx_moves_account_balance index on Moves table.

flemzord committed:
docs: add notes.yaml files for split index migrations

Add comprehensive notes.yaml documentation for each migration:
- Migration 41: idx_moves_pit_insertion index creation
- Migration 42: idx_moves_pit_effective index creation
- Migration 43: idx_accounts_metadata_pit index creation
- Migration 44: idx_transactions_metadata_pit index creation
- Migration 45: Drop all old indexes

Each file documents the purpose, changes, affected queries,
performance impact, and safety considerations.

flemzord committed:
fix: remove CONCURRENTLY from index creation migrations

CONCURRENTLY cannot be used within the migration framework as migrations
run within a transactional context (advisory locks). Removed CONCURRENTLY
from all index creation statements to ensure migrations can execute successfully.

The migrations remain isolated (one index per migration) which maintains
good organization and allows for easier rollback if needed. Index creation
will lock tables during execution but ensures compatibility with the
migration system.

Note: Migration 41-44 create new indexes, migration 45 drops old ones.
This approach minimizes risk by creating all new indexes before dropping
old ones.

flemzord committed:
fix: correct column name from account_address to accounts_address in index creation

flemzord committed:
fix: rename testServer variable to srv for clarity and add cleanup on service stop

flemzord committed:
Increase default benchmark count from 1 to 5

flemzord committed:
feat: add index to optimize update_effective_volumes trigger

Add idx_moves_update_effective_volumes index to prevent sequential scans
on every INSERT when MOVES_HISTORY feature is enabled (ON by default).

The update_effective_volumes trigger runs after each move insertion and
updates all moves with effective_date > new.effective_date. Without this
index, this operation causes significant performance degradation at scale.

Changes:
- New migration 45: Create idx_moves_update_effective_volumes
- Migration 45 renamed to 46: Updated to drop moves_range_dates

flemzord committed:
fix: Add concurrently

flemzord committed:
fix: migrations
